qiskit/transpiler/passes/unroller.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""Pass for unrolling a circuit to a given basis."""
+
+from qiskit.transpiler._basepasses import TransformationPass
+from qiskit.exceptions import QiskitError
+
+
+class Unroller(TransformationPass):
+    """
+    Unroll (expand) non-basis, non-opaque instructions recursively
+    to a desired basis, using decomposition rules defined for each instruction.
+    """
+
+    def __init__(self, basis):
+        """
+        Args:
+            basis (list[str]): Target basis gate names to unroll to, e.g. `['U', 'CX']` .
+        """
+        super().__init__()
+        self.basis = basis
+
+    def run(self, dag):
+        """Expand all op nodes to the given basis.
+
+        If self.basis is empty, the circuit is unrolled down to
+        fundamental (opaque) gates (U, CX).
+
+        Args:
+            dag(DAGCircuit): input dag
+
+        Raises:
+            QiskitError: if unable to unroll given the basis due to undefined
+            decomposition rules (such as a bad basis) or excessive recursion.
+
+        Returns:
+            DAGCircuit: output unrolled dag
+        """
+        # Walk through the DAG and expand each non-basis node
+        for node in dag.gate_nodes():
+            current_node = dag.multi_graph.node[node]
+            if current_node["op"].name in self.basis:  # If already a base, ignore.
+                continue
+
+            # TODO: allow choosing other possible decompositions
+            try:
+                decomposition_rules = current_node["op"].decompositions()
+            except NotImplementedError:
+                raise QiskitError("Cannot unroll the circuit to the given basis, "
+                                  + str(self.basis) +
+                                  ". No decomposition rules defined for "
+                                  + current_node["op"].name + ".")
+            except RecursionError:
```

Comment:
Let's just remove ReursionError for now, and we'll add it if need be later. I see how you have checked for it, but I'm not sure this is the right place. I don't think it will be raised in your `try` block.